core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt

```diff
@@ -0,0 +1,206 @@
+    if (result.type == HitTestResult.IMAGE_ANCHOR_TYPE ||
+      result.type == HitTestResult.IMAGE_TYPE ||
+      result.type == HitTestResult.SRC_IMAGE_ANCHOR_TYPE
+    ) {
+      val saveMenu =
+        menu.add(0, 1, 0, resources.getString(R.string.save_media))
+      saveMenu.setOnMenuItemClickListener {
+        val msg = SaveHandler(zimReaderContainer).obtainMessage()
+        requestFocusNodeHref(msg)
+        true
+      }
+    }
+  }
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    compositeDisposable.add(
+      sharedPreferenceUtil.textZooms
+        .subscribe {
+          settings.textZoom = it
+        }
+    )
+  }
+
+  override fun onDetachedFromWindow() {
+    super.onDetachedFromWindow()
+    compositeDisposable.clear()
+  }
+
+  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
+    super.onScrollChanged(l, t, oldl, oldt)
+    val windowHeight = if (measuredHeight > 0) measuredHeight else 1
+    val pages = contentHeight / windowHeight
+    val page = t / windowHeight
+    callback.webViewPageChanged(page, pages)
+  }
+
+  internal class SaveHandler(private val zimReaderContainer: ZimReaderContainer) :
+    Handler() {
+    private fun getDecodedFileName(url: String?, src: String?): String {
+      var fileName = ""
+      if (url != null) {
+        fileName = url.substringAfterLast("/")
+      }
+      // If url is not a valid file name use src if it isn't null
+      if (!fileName.contains(".") && src != null) {
+        fileName = src.substringAfterLast("/")
+      }
+      return fileName.substringAfterLast("%3A")
+    }
+
+    override fun handleMessage(msg: Message) {
+      val url = msg.data["url"] as? String
+      val src = msg.data["src"] as? String
+      if (url != null || src != null) {
+        val fileName = getDecodedFileName(url, src)
+        val dotIndex = fileName.lastIndexOf('.')
+        var root =
+          Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
+        if (instance.externalMediaDirs.isNotEmpty()) {
+          root = instance.externalMediaDirs[0]
+        }
+        var storageDir = File(root, fileName)
+        var newUrl = fileName
+        var i = 2
+        while (storageDir.exists()) {
+          newUrl = fileName.substring(0, dotIndex) + "_" + i + fileName.substring(dotIndex)
+          storageDir = File(root, newUrl)
+          i++
+        }
+        val source = Uri.parse(src)
+        try {
+          val input = zimReaderContainer.load("$source").data
+          storageDir.writeText(input.readBytes().toString())
+          val imageSaved = instance.getString(R.string.save_media_saved)
+          val toastText = String.format(imageSaved, newUrl)
+          instance.toast(toastText)
+        } catch (e: IOException) {
+          Log.w("kiwix", "Couldn't save image", e)
+          val toastText = instance.getString(R.string.save_media_error)
+          instance.toast(toastText)
```

Comment:
we don't need toastText if the preceding string is just a call to `getString`, our toast API accepts StringIds

Reply:
done